src/greenlight_alby_client.rs

```diff
@@ -0,0 +1,217 @@
+            description: req.description,
+            ..Default::default()
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct GreenlightInvoiceResponse {
+    pub bolt11: String,
+}
+
+impl From<cln::InvoiceResponse> for GreenlightInvoiceResponse {
+    fn from(invoice: cln::InvoiceResponse) -> Self {
+        GreenlightInvoiceResponse {
+            bolt11: invoice.bolt11,
+        }
+    }
+}
+
+pub struct GreenlightAlbyClient {
+    // signer: gl_client::signer::Signer,
+    scheduler: gl_client::scheduler::Scheduler,
+    tls: TlsConfig,
+}
+
+pub async fn recover(mnemonic: String) -> Result<GreenlightCredentials> {
+    let mnemonic = Mnemonic::from_str(&mnemonic)
+        .context("failed to parse mnemonic")
+        .map_err(SdkError::invalid_arg)?;
+
+    let secret = mnemonic.to_seed("")[0..32].to_vec(); // Only need the first 32 bytes
+
+    let tls = TlsConfig::new()
+        .context("failed to create TLS config")
+        .map_err(SdkError::greenlight_api)?;
+
+    let signer = Signer::new(secret, Network::Bitcoin, tls)
+        .context("failed to create signer")
+        .map_err(SdkError::greenlight_api)?;
+
+    let scheduler = Scheduler::new(signer.node_id(), Network::Bitcoin)
+        .await
+        .context("failed to create scheduler")
+        .map_err(SdkError::greenlight_api)?;
+
+    Ok(scheduler
+        .recover(&signer)
+        .await
+        .context("failed to recover credentials")
+        .map_err(SdkError::greenlight_api)?
+        .into())
+}
+
+pub async fn new_greenlight_alby_client(
+    mnemonic: String,
+    credentials: GreenlightCredentials,
+) -> Result<Arc<GreenlightAlbyClient>> {
+    let tls = TlsConfig::new()
+        .context("failed to create TLS config")
+        .map_err(SdkError::greenlight_api)?
+        .identity(
+            credentials.device_cert.into_bytes(),
+            credentials.device_key.into_bytes(),
+        );
+
+    let mnemonic = Mnemonic::from_str(&mnemonic)
+        .context("failed to parse mnemonic")
+        .map_err(SdkError::invalid_arg)?;
+
+    let secret = mnemonic.to_seed("")[0..32].to_vec(); // Only need the first 32 bytes
+
+    let signer = Signer::new(secret, Network::Bitcoin, tls.clone())
+        .context("failed to create signer")
+        .map_err(SdkError::greenlight_api)?;
+    let scheduler = Scheduler::new(signer.node_id(), Network::Bitcoin)
+        .await
+        .context("failed to create scheduler")
+        .map_err(SdkError::greenlight_api)?;
+
+    Ok(Arc::new(GreenlightAlbyClient {
+        tls,
+        scheduler,
+        // signer,
+    }))
+}
+
+impl GreenlightAlbyClient {
+    async fn get_node(&self) -> Result<gl_client::node::ClnClient> {
+        // wakes up the node
+        self.scheduler
+            .schedule(self.tls.clone())
+            .await
+            .context("failed to schedule node")
+            .map_err(SdkError::greenlight_api)
+    }
+
+    pub async fn get_info(&self) -> Result<GreenlightNodeInfo> {
```

Comment:
Can we make this return a pointer if there is an error? So in go the value would be nil

Reply:
We can, but then we would need to take care of adding a deallocator method and deallocating the memory manually, which is a bit of PITA. However, this translates into an idiomatic Go method returning `(GreenlightNodeInfo, error)`.

Reply:
Disregard that — I realized that uniffi supports Arc for that. Will fix pronto

Reply:
No, wait. Apologies for the noise! That's only for interfaces, unfortunately. We can define this type as an interface (instead of a dictionary) in the udl file, but then we'll need to add getters for the member fields and use them to access data. If it's worth it, I shall update the code.

Reply:
ok, maybe we just leave it as is. Thanks for looking into it!